R/card.R

```diff
@@ -0,0 +1,298 @@
+#'   vertical space is available in the card.
+#'
+#' @export
+card_body <- function(..., class = NULL, padding = c("x", "y"), stretch = FALSE) {
+  res <- htmltools::tags$div(class = "card-body",
+    class = if (!"x" %in% padding) "px-0",
+    class = if (!"y" %in% padding) "py-0",
+    class = class,
+    ...,
+    style = if (!stretch) css(
+      flex = "0",
+      `_webkit_flex` = "0"
+    ),
+  )
+  as.card_item(res)
+}
+
+#' @export
+card_body_scroll <- function(..., class = NULL, height = NULL, padding = c("x", "y")) {
+  res <- htmltools::tags$div(class = "card-body card-body-scroll",
+    class = if (!"x" %in% padding) "px-0",
+    class = if (!"y" %in% padding) "py-0",
+    class = class,
+    style = css(
+      # May be NULL
+      flex_basis = validateCssUnit(height),
+      `-webkit-flex-basis` = validateCssUnit(height)
+    ),
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_header <- function(..., class = NULL) {
+  res <- htmltools::tags$div(class = "card-header", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_heading <- function(..., class = NULL, container = htmltools::tags$h6) {
+  res <- container(class = "card-header mt-0", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_title <- function(..., class = NULL) {
+  res <- htmltools::tags$h5(class = "card-header", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_footer <- function(..., class = NULL) {
+  res <- htmltools::tags$div(class = "card-footer mt-auto", class = class,
+    ...
+  )
+  as.card_item(res)
+}
+
+#' @rdname card_body
+#' @export
+card_spacer <- function(...) {
+  res <- htmltools::tags$div(style = css(flex = "1 0 0"))
+  as.card_item(res)
+}
+
+# jcheng 2022-06-06: Removing for now; list items have more features than I'm
+# ready to design an API for right now
+#
+# #' @rdname card_body
+# #' @export
+# card_list <- function(...) {
+#   res <- htmltools::tags$ul(class = "list-group list-group-flush", ...)
+#   as.card_item(res)
+# }
+#
+# #' @export
+# card_list_item <- function(...) {
+#   htmltools::tags$li(class = "list-group-item", ...)
+# }
+
+#' @rdname card_body
+#' @export
+card_plot_output <- function(outputId,
+  click = NULL,
+  dblclick = NULL,
+  hover = NULL,
+  brush = NULL,
+  height = NULL,
+  stretch = TRUE,
+  ...
+) {
+  plot_div <- plotOutput(outputId, height = NULL, click = click, dblclick = dblclick, hover = hover,
+    brush = brush)
+
+  # TODO: card-img-* needs to go on the <img> itself, not the containing <div>
+  plot_div <-
+    tagAppendAttributes(plot_div,
+      style = css(
+        flex = if (stretch) "1 1",
+        `-webkit-flex` = if (stretch) "1 1",
+        # May be NULL
+        `flex-basis` = validateCssUnit(height),
+        `-webkit-flex-basis` = validateCssUnit(height),
```

Comment:
I find the two-value syntax confusing since the interpretation of the 2nd number changes depending on the value's unit https://developer.mozilla.org/en-US/docs/Web/CSS/flex#syntax

What made you add the `-webkit-*` prefix? Is it only because `bs_theme()` adds them? Since those are added only for phantomjs (`{shinytest}`) support, I'd actually prefer that we don't add them here (more motivation to migrate to `{shinytest2}`)


```suggestion
        flex_grow = if (stretch) "1",
        flex_shrink = if (stretch) "1",
        # May be NULL
        flex_basis = validateCssUnit(height),
```